File: src/caeluviim/canonical.py
from __future__ import annotations

import base64
import hashlib
import json
import math
import unicodedata
from copy import deepcopy
from typing import Any, Mapping

from cryptography.hazmat.primitives import serialization
from cryptography.hazmat.primitives.asymmetric.ed25519 import (
    Ed25519PrivateKey,
    Ed25519PublicKey,
)

BASE58_ALPHABET = "123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz"
MAX_SAFE_INTEGER = 9_007_199_254_740_991


class CanonicalizationError(ValueError):
    pass


def _normalize(value: Any, stack: set[int]) -> Any:
    if value is None or isinstance(value, bool):
        return value
    if isinstance(value, str):
        return unicodedata.normalize("NFC", value)
    if isinstance(value, int):
        if abs(value) > MAX_SAFE_INTEGER:
            raise CanonicalizationError("canonical JSON permits safe integers only")
        return value
    if isinstance(value, float):
        raise CanonicalizationError("canonical JSON forbids binary floating point")
    if isinstance(value, (bytes, bytearray, memoryview)):
        raise CanonicalizationError("binary values must be stored as content-addressed objects")

    marker = id(value)
    if marker in stack:
        raise CanonicalizationError("cyclic input is not canonical JSON")
    stack.add(marker)
    try:
        if isinstance(value, (list, tuple)):
            return [_normalize(item, stack) for item in value]
        if isinstance(value, Mapping):
            normalized: dict[str, Any] = {}
            for key, item in value.items():
                if not isinstance(key, str):
                    raise CanonicalizationError("canonical JSON object keys must be strings")
                normalized_key = unicodedata.normalize("NFC", key)
                if normalized_key in normalized:
                    raise CanonicalizationError("object keys collide after NFC normalization")
                normalized[normalized_key] = _normalize(item, stack)
            return dict(
                sorted(normalized.items(), key=lambda pair: pair[0].encode("utf-8"))
            )
        raise CanonicalizationError(
            f"unsupported canonical JSON value: {type(value).__name__}"
        )
    finally:
        stack.remove(marker)


def canonical_encode(value: Any) -> str:
    normalized = _normalize(value, set())
    return json.dumps(
        normalized,
        ensure_ascii=False,
        separators=(",", ":"),
        allow_nan=False,
    )


def canonical_bytes(value: Any) -> bytes:
    return canonical_encode(value).encode("utf-8")


def sha256_hex(data: bytes) -> str:
    return hashlib.sha256(data).hexdigest()


def domain_digest(domain: str, value: Any) -> bytes:
    return hashlib.sha256(domain.encode("utf-8") + b"\x00" + canonical_bytes(value)).digest()


def base58btc(data: bytes) -> str:
    leading_zeroes = len(data) - len(data.lstrip(b"\x00"))
    number = int.from_bytes(data, "big")
    encoded = ""
    while number:
        number, remainder = divmod(number, 58)
        encoded = BASE58_ALPHABET[remainder] + encoded
    return "z" + ("1" * leading_zeroes) + encoded


def content_urn(kind: str, digest_hex: str) -> str:
    return f"urn:caeluviim:{kind}:sha256:{digest_hex}"


def operation_body(operation: Mapping[str, Any]) -> dict[str, Any]:
    body = deepcopy(dict(operation))
    body.pop("operation_id", None)
    body.pop("content_hash", None)
    body.pop("signature", None)
    return body


def _validate_utf8_string_set(values: Any, label: str) -> None:
    if not isinstance(values, list) or not all(isinstance(value, str) for value in values):
        raise CanonicalizationError(f"{label} must be a string set")
    sorted_values = sorted(values, key=lambda value: value.encode("utf-8"))
    if len(set(values)) != len(values) or values != sorted_values:
        raise CanonicalizationError(f"{label} must be unique and UTF-8 sorted")


def operation_digest(operation: Mapping[str, Any]) -> bytes:
    authorization = operation.get("authorization")
    authority_ids = authorization.get("authority_ids") if isinstance(authorization, Mapping) else None
    for values, label in (
        (operation.get("evidence_ids"), "evidence_ids"),
        (operation.get("parent_ids"), "parent_ids"),
        (operation.get("dependencies"), "dependencies"),
        (authority_ids, "authorization.authority_ids"),
    ):
        _validate_utf8_string_set(values, label)
    return domain_digest("DAP-OPERATION-0.2", operation_body(operation))


def operation_identifiers(operation: Mapping[str, Any]) -> tuple[str, str, bytes]:
    digest = operation_digest(operation)
    return (
        f"op:{base58btc(digest)}",
        f"sha256:{digest.hex()}",
        digest,
    )


def ruleset_identifier(ruleset: Mapping[str, Any]) -> str:
    body = deepcopy(dict(ruleset))
    body.pop("ruleset_id", None)
    return f"ruleset:{base58btc(domain_digest('DAP-RULESET-0.2', body))}"


def generate_ed25519_private_key() -> Ed25519PrivateKey:
    return Ed25519PrivateKey.generate()


def private_key_bytes(key: Ed25519PrivateKey) -> bytes:
    return key.private_bytes(
        encoding=serialization.Encoding.PEM,
        format=serialization.PrivateFormat.PKCS8,
        encryption_algorithm=serialization.NoEncryption(),
    )


def public_key_bytes(key: Ed25519PublicKey) -> bytes:
    return key.public_bytes(
        encoding=serialization.Encoding.Raw,
        format=serialization.PublicFormat.Raw,
    )


def load_private_key(data: bytes) -> Ed25519PrivateKey:
    key = serialization.load_pem_private_key(data, password=None)
    if not isinstance(key, Ed25519PrivateKey):
        raise ValueError("expected an Ed25519 private key")
    return key


def sign_digest(key: Ed25519PrivateKey, digest: bytes) -> str:
    return base64.urlsafe_b64encode(key.sign(digest)).rstrip(b"=").decode("ascii")


def verify_digest(key: Ed25519PublicKey, digest: bytes, signature: str) -> bool:
    padding = "=" * ((4 - len(signature) % 4) % 4)
    try:
        key.verify(base64.urlsafe_b64decode(signature + padding), digest)
        return True
    except Exception:
        return False

File: web/lib/dap/canonical.ts
import type { DapOperationEnvelope, DapRuleset } from "./types";

const encoder = new TextEncoder();
const base58Alphabet = "123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz";

function compareUtf8(left: string, right: string) {
  const a = encoder.encode(left);
  const b = encoder.encode(right);
  const length = Math.min(a.length, b.length);
  for (let index = 0; index < length; index += 1) {
    if (a[index] !== b[index]) return a[index] - b[index];
  }
  return a.length - b.length;
}

function encode(value: unknown, stack: Set<object>): string {
  if (value === null || typeof value === "boolean") return JSON.stringify(value);
  if (typeof value === "string") return JSON.stringify(value.normalize("NFC"));
  if (typeof value === "number") {
    if (!Number.isSafeInteger(value) || Object.is(value, -0)) {
      throw new TypeError("DAP canonical JSON permits safe integers only");
    }
    return String(value);
  }
  if (typeof value !== "object" || value === undefined) {
    throw new TypeError(`Unsupported DAP canonical JSON value: ${typeof value}`);
  }
  const prototype = Object.getPrototypeOf(value);
  if (!Array.isArray(value) && prototype !== Object.prototype && prototype !== null) {
    throw new TypeError("DAP canonical JSON objects must be plain records");
  }
  if (stack.has(value)) throw new TypeError("Cyclic input is not canonical JSON");
  stack.add(value);
  try {
    if (Array.isArray(value)) {
      return `[${value.map((item) => encode(item, stack)).join(",")}]`;
    }
    const normalized = Object.entries(value).map(([key, item]) => [key.normalize("NFC"), item] as const);
    if (new Set(normalized.map(([key]) => key)).size !== normalized.length) {
      throw new TypeError("Object keys collide after NFC normalization");
    }
    normalized.sort(([left], [right]) => compareUtf8(left, right));
    return `{${normalized.map(([key, item]) => `${JSON.stringify(key)}:${encode(item, stack)}`).join(",")}}`;
  } finally {
    stack.delete(value);
  }
}

export function canonicalEncode(value: unknown) {
  return encode(value, new Set());
}

export function canonicalBytes(value: unknown) {
  return encoder.encode(canonicalEncode(value));
}

function bytesToHex(bytes: Uint8Array) {
  return [...bytes].map((byte) => byte.toString(16).padStart(2, "0")).join("");
}

export function base64urlToBytes(value: string) {
  const normalized = value.replaceAll("-", "+").replaceAll("_", "/");
  const padded = `${normalized}${"=".repeat((4 - (normalized.length % 4)) % 4)}`;
  const binary = atob(padded);
  return Uint8Array.from(binary, (character) => character.charCodeAt(0));
}

export function bytesToBase64url(bytes: Uint8Array) {
  let binary = "";
  for (const byte of bytes) binary += String.fromCharCode(byte);
  return btoa(binary).replaceAll("+", "-").replaceAll("/", "_").replaceAll("=", "");
}

export function base58btc(bytes: Uint8Array) {
  let leadingZeroes = 0;
  while (leadingZeroes < bytes.length && bytes[leadingZeroes] === 0) leadingZeroes += 1;
  let number = BigInt(0);
  for (const byte of bytes) number = number * BigInt(256) + BigInt(byte);
  let encoded = "";
  while (number > BigInt(0)) {
    encoded = `${base58Alphabet[Number(number % BigInt(58))]}${encoded}`;
    number /= BigInt(58);
  }
  return `z${"1".repeat(leadingZeroes)}${encoded}`;
}

export function assertSortedStringSet(values: unknown, label: string) {
  if (!Array.isArray(values) || values.some((value) => typeof value !== "string")) {
    throw new TypeError(`${label} must be a string set`);
  }
  const sorted = [...values].sort(compareUtf8);
  if (new Set(values).size !== values.length || values.some((value, index) => value !== sorted[index])) {
    throw new TypeError(`${label} must be unique and UTF-8 sorted`);
  }
}

export async function domainDigest(domain: string, value: unknown) {
  const prefix = encoder.encode(domain);
  const body = canonicalBytes(value);
  const input = new Uint8Array(prefix.length + 1 + body.length);
  input.set(prefix, 0);
  input[prefix.length] = 0;
  input.set(body, prefix.length + 1);
  return new Uint8Array(await crypto.subtle.digest("SHA-256", input));
}

export function operationBody(operation: DapOperationEnvelope | Record<string, unknown>) {
  const body = { ...operation } as Record<string, unknown>;
  delete body.operation_id;
  delete body.content_hash;
  delete body.signature;
  return body;
}

export async function operationIdentifiers(operation: DapOperationEnvelope | Record<string, unknown>) {
  const candidate = operation as Partial<DapOperationEnvelope>;
  assertSortedStringSet(candidate.evidence_ids, "evidence_ids");
  assertSortedStringSet(candidate.parent_ids, "parent_ids");
  assertSortedStringSet(candidate.dependencies, "dependencies");
  assertSortedStringSet(candidate.authorization?.authority_ids, "authorization.authority_ids");
  const digest = await domainDigest("DAP-OPERATION-0.2", operationBody(operation));
  return {
    digest,
    operationId: `op:${base58btc(digest)}`,
    contentHash: `sha256:${bytesToHex(digest)}`,
  };
}

export function rulesetBody(ruleset: DapRuleset | Record<string, unknown>) {
  const body = { ...ruleset } as Record<string, unknown>;
  delete body.ruleset_id;
  return body;
}

export async function rulesetIdentifier(ruleset: DapRuleset | Record<string, unknown>) {
  const digest = await domainDigest("DAP-RULESET-0.2", rulesetBody(ruleset));
  return `ruleset:${base58btc(digest)}`;
}

export async function stateRoot(state: unknown) {
  const digest = await domainDigest("DAP-STATE-0.2", state);
  return `state:${base58btc(digest)}`;
}

export async function historyRoot(operationIds: string[]) {
  const digest = await domainDigest("DAP-HISTORY-0.2", operationIds);
  return `history:${base58btc(digest)}`;
}

export async function verifyEd25519(publicKeyBase64url: string, signatureBase64url: string, digest: Uint8Array) {
  try {
    const publicKey = base64urlToBytes(publicKeyBase64url);
    const signature = base64urlToBytes(signatureBase64url);
    if (publicKey.length !== 32 || signature.length !== 64) return false;
    const key = await crypto.subtle.importKey("raw", publicKey, { name: "Ed25519" }, false, ["verify"]);
    return crypto.subtle.verify(
      { name: "Ed25519" },
      key,
      signature.slice().buffer as ArrayBuffer,
      digest.slice().buffer as ArrayBuffer,
    );
  } catch {
    return false;
  }
}

export async function signEd25519Pkcs8(privateKeyPkcs8Base64url: string, digest: Uint8Array) {
  const privateKey = base64urlToBytes(privateKeyPkcs8Base64url);
  const key = await crypto.subtle.importKey("pkcs8", privateKey, { name: "Ed25519" }, false, ["sign"]);
  const signature = await crypto.subtle.sign(
    { name: "Ed25519" },
    key,
    digest.slice().buffer as ArrayBuffer,
  );
  return bytesToBase64url(new Uint8Array(signature));
}
